components/script/window_named_properties.rs

```diff
@@ -0,0 +1,206 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::dom::bindings::codegen::Bindings::WindowBinding::WindowMethods;
+use crate::dom::bindings::proxyhandler::fill_property_descriptor;
+use crate::dom::bindings::root::Root;
+use crate::dom::bindings::utils::has_property_on_prototype;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::window::Window;
+use crate::js::conversions::ToJSValConvertible;
+use crate::script_runtime::JSContext as SafeJSContext;
+use js::conversions::jsstr_to_string;
+use js::error::throw_type_error;
+use js::glue::RUST_JSID_TO_STRING;
+use js::glue::{CreateProxyHandler, NewProxyObject, ProxyTraps, RUST_JSID_IS_STRING};
+use js::jsapi::JS_SetImmutablePrototype;
+use js::jsapi::{Handle, HandleObject, JSClass, JSContext, JSErrNum, UndefinedHandleValue};
+use js::jsapi::{
+    HandleId, JSClass_NON_NATIVE, MutableHandle, ObjectOpResult, PropertyDescriptor,
+    ProxyClassExtension, ProxyClassOps, ProxyObjectOps, JSCLASS_DELAY_METADATA_BUILDER,
+    JSCLASS_IS_PROXY, JSCLASS_RESERVED_SLOTS_MASK, JSCLASS_RESERVED_SLOTS_SHIFT,
+};
+use js::jsval::UndefinedValue;
+use js::rust::IntoHandle;
+use js::rust::{Handle as RustHandle, MutableHandle as RustMutableHandle};
+use js::rust::{HandleObject as RustHandleObject, MutableHandleObject as RustMutableHandleObject};
+use libc;
+use std::ptr;
+
+struct SyncWrapper(*const libc::c_void);
+#[allow(unsafe_code)]
+unsafe impl Sync for SyncWrapper {}
+
+lazy_static! {
+    static ref HANDLER: SyncWrapper = {
+        let traps = ProxyTraps {
+            enter: None,
+            getOwnPropertyDescriptor: Some(get_own_property_descriptor),
+            defineProperty: Some(define_property),
+            ownPropertyKeys: None,
```

Comment:
For reference, the corresponding function in Gecko is https://searchfox.org/mozilla-central/source/dom/base/WindowNamedPropertiesHandler.cpp#168.